chiselFrontend/src/main/scala/chisel3/core/Module.scala

```diff
@@ -79,34 +71,122 @@ object Module {
-    iodef
+  Builder.currentModule = Some(this)
+  Builder.whenDepth = 0
+
+  //
+  // Module Construction Internals
+  //
+  protected var _closed = false
+
+  // Fresh Namespace because in Firrtl, Modules namespaces are disjoint with the global namespace
+  private[core] val _namespace = Namespace.empty
+  protected val _ids = ArrayBuffer[HasId]()
+  private[chisel3] def addId(d: HasId) {
+    require(!_closed, "Can't write to module after module close")
+    _ids += d
   }
 
-  def annotate(annotation: ChiselAnnotation): Unit = {
-    Builder.annotations += annotation
+  protected val _ports = new ArrayBuffer[Data]()
+
+  /** Generates the FIRRTL Component (Module or Blackbox) of this Module.
+    * Also closes the module so no more construction can happen inside.
+    */
+  private[core] def generateComponent(): Component
+
+  /** Sets up this module in the parent context
+    */
+  private[core] def initializeInParent()
+
+  //
+  // Chisel Internals
+  //
+  /** Desired name of this module. Override this to give this module a custom, perhaps parametric,
+    * name.
+    */
+  def desiredName = this.getClass.getName.split('.').last
+
+  /** Legalized name of this module. */
+  final val name = Builder.globalNamespace.name(desiredName)
+
+  /** Called at the Module.apply(...) level after this Module has finished elaborating.
+    * Returns a map of nodes -> names, for named nodes.
+    *
+    * Helper method.
+    */
+  protected def nameIds(rootClass: Class[_]): HashMap[HasId, String] = {
+    val names = new HashMap[HasId, String]()
+
+    def name(node: HasId, name: String) {
+      // First name takes priority, like suggestName
+      // TODO: DRYify with suggestName
+      if (!names.contains(node)) {
+        names.put(node, name)
+      }
+    }
+
+    /** Recursively suggests names to supported "container" classes
+      * Arbitrary nestings of supported classes are allowed so long as the
+      * innermost element is of type HasId
+      * (Note: Map is Iterable[Tuple2[_,_]] and thus excluded)
+      */
+    def nameRecursively(prefix: String, nameMe: Any): Unit =
+      nameMe match {
+        case (id: HasId) => name(id, prefix)
+        case Some(elt) => nameRecursively(prefix, elt)
+        case (iter: Iterable[_]) if iter.hasDefiniteSize =>
+          for ((elt, i) <- iter.zipWithIndex) {
+            nameRecursively(s"${prefix}_${i}", elt)
+          }
+        case _ => // Do nothing
+      }
+
+    /** Scala generates names like chisel3$util$Queue$$ram for private vals
+      * This extracts the part after $$ for names like this and leaves names
+      * without $$ unchanged
+      */
+    def cleanName(name: String): String = name.split("""\$\$""").lastOption.getOrElse(name)
+
+    for (m <- getPublicFields(rootClass)) {
+      nameRecursively(cleanName(m.getName), m.invoke(this))
+    }
+
+    // For Module instances we haven't named, suggest the name of the Module
+    _ids foreach {
```

Comment:
Why is backup naming of modules done here? This seems like it should be done when registers, wires, etc. also need backup naming.